src/utils.c

```diff
@@ -92,6 +92,33 @@ SEXP vctrs_dispatch3(SEXP fn_sym, SEXP fn,
   return out;
 }
 
+SEXP df_map(SEXP df, SEXP (*fn)(SEXP)) {
+  R_len_t n = Rf_length(df);
+  SEXP out = PROTECT(Rf_allocVector(VECSXP, n));
+
+  for (R_len_t i = 0; i < n; ++i) {
+    SET_VECTOR_ELT(out, i, fn(VECTOR_ELT(df, i)));
+  }
+
+  // FIXME: Should that be restored?
+  SEXP nms = PROTECT(Rf_getAttrib(df, R_NamesSymbol));
+  Rf_setAttrib(out, R_NamesSymbol, nms);
+
+  out = df_restore(out, df, vctrs_shared_empty_int);
+
+  UNPROTECT(2);
+  return out;
+}
+
+SEXP vec_recurse(SEXP x, SEXP (*rec)(SEXP, bool), SEXP i) {
```

Comment:
Why is this called `vec_recurse()`?

Reply:
Because it recurses into `rec` with `dispatch = false`.

Reply:
What do you mean by recurses into? Where is the recursion?

Reply:
I mean `vec_method(dispatch = true) => vec_recurse() => vec_method(dispatch = false)`

Reply:
Yeah, I just don't find it to be a particularly evocative name

Reply:
maybe `vec_dispatch()`?